src/tls13.c

```diff
@@ -7180,40 +7180,42 @@ int DoTls13HandShakeMsgType(WOLFSSL* ssl, byte* input, word32* inOutIdx,
         /* Need to hash input message before deriving secrets. */
     #ifndef NO_WOLFSSL_CLIENT
         if (ssl->options.side == WOLFSSL_CLIENT_END) {
-            if (type == server_hello) {
-                if ((ret = DeriveEarlySecret(ssl)) != 0)
-                    return ret;
-                if ((ret = DeriveHandshakeSecret(ssl)) != 0)
-                    return ret;
-
-                if ((ret = DeriveTls13Keys(ssl, handshake_key,
-                                           ENCRYPT_AND_DECRYPT_SIDE, 1)) != 0) {
-                    return ret;
+            if (ssl->options.handShakeState != CLIENT_HELLO_COMPLETE) {
```

Comment:
Why is this necessary?
The ret is checked for 0 and the message type is server_hello or finished.

Reply:
I made this change to prevent having to re-calculate this on each re-entry during the re-handshake. Let me test without it and see if it's really being called multiple. Now that you point me to it I think the `type` will do that properly.